Dask.order rewrite using a critical path approach

fjetter opened the pull request:
This is another ~attempt to~ rewrite of dask.order

Supersedes https://github.com/dask/dask/pull/10557
Closes https://github.com/dask/distributed/issues/8255


It does not incorporate https://github.com/dask/dask/pull/10619 yet but I will likely want to follow up adding this

This takes a quite different approach to our ordering that the PRs before this and actually rewrites it from scratch.

I still have a couple of unit tests to take care of but the failing tests right now are mostly artificial test cases. There are one or two mildly concerning things I still have to look at but generally speaking I'm optimistic about this approach

High level, this change was inspired by what the root task queuing is doing in that it tries to finish off whatever runnable computation branches exist before it starts loading new stuff. On such a high level, the algorithm can be described in a couple of simple steps

1. Compute a critical path by reverse DFS
2. Use the critical path as a stack and walk the stack in forward direction. Whenever a node is encountered, that is not runnable, add the missing dependencies
3. Whenever that happens, look at all currently runnable tasks and walk them as far as possible. Those paths are remembered but not immediately added to the result
4. If the runnable paths meet a reducer (or final node) and that reducer would be reachable given all runnable paths, execute those paths
5. If not, proceed with the critical path until all nodes are processed


The nuances of the implementation are a little harder to describe but I'll add some more in-code comments.

fjetter commented:
coiled benchmark results are in

https://github.com/coiled/benchmarks/actions/runs/7046212948

Wall clock time is mostly unaffected. The little signal there is likely just noise

Looking at peak memory, we see a reduction for double_diff and dataframe_align

![image](https://github.com/dask/dask/assets/8629629/74fc3c55-db47-4dd6-9cb6-29bd08996f79)

and a very solid reduction for the zarr filtering tests

![image](https://github.com/dask/dask/assets/8629629/1358a145-5931-4988-b8b0-8bd928b47b07)

Similar patterns can be seen in average memory use

fjetter commented:
`doublediff` does not drop as strongly as in https://github.com/dask/dask/pull/10557 so there is likely still room for improvement. Given that the cost function / sorting key I am using is almost random, I assume this is the most interesting knob I have to tweak

fjetter commented:
This critical path approach has one flaw I haven't quite figured out, yet. 

The following example is a smaller version of what was reported in https://github.com/dask/dask/issues/3055

![image](https://github.com/dask/dask/assets/8629629/9ce73fdc-3f6d-4620-a435-dd092a549ef3)

The visualization shows clearly that we're very eagerly processing the critical path and are working off unlocked branches as soon as they are available. However, with this particular graph, a root task that is required to reduce everything early on is only required by the _critical_ path very late on such that those branches never get a chance to run

fjetter commented:
So, I changed how the critical path is determined such that we have two different strategies based on a heuristic about graph symmetry. That is working pretty well for the test suite and the benchmark suite.

https://github.com/coiled/benchmarks/actions/runs/7088167915

Wall time looks great for all H20 tests and the zarr filtering

<img width="587" alt="Screenshot 2023-12-05 at 09 25 38" src="https://github.com/dask/dask/assets/8629629/48a90e3b-cce5-430d-ad4c-037b47e543e9">
<img width="619" alt="Screenshot 2023-12-05 at 09 25 59" src="https://github.com/dask/dask/assets/8629629/37ab208a-55e1-418c-b2c9-e411fad6e433">

There appears to be a mild regression in the shuffle tests I still want to investigate since I'm surprised those are affected at all (maybe this is runtime of the algo, idk)

<img width="411" alt="Screenshot 2023-12-05 at 09 26 17" src="https://github.com/dask/dask/assets/8629629/ff739694-77a9-450b-916d-02350c786e29">


Memory looks great overall (below peak memory, avg looks similar)

<img width="617" alt="Screenshot 2023-12-05 at 09 27 03" src="https://github.com/dask/dask/assets/8629629/7a6e9452-c0df-4f22-8f00-17ababf1bdc7">
<img width="567" alt="Screenshot 2023-12-05 at 09 27 27" src="https://github.com/dask/dask/assets/8629629/c3632f4b-3e12-4cee-8409-e51f50abff6a">

mrocklin commented:
I'm mildly curious when we'll be able to merge and release this.  There's probably some mild benefit to it going out before AGU (next week) if that's doable.

mrocklin commented:
cc @dcherian

fjetter commented:
I reran the benchmarks and the shuffle related regressions that showed up earlier did vanish. The improvements in H20 unfortunately also vanished but that's somewhat expected. 

https://github.com/coiled/benchmarks/actions/runs/7130181112

Note, that this rewrite does not include the "skip data tasks" as proposed inhttps://github.com/dask/dask/pull/10619 yet but it does fix the test case I added there to close https://github.com/dask/dask/issues/10618 so I added it also to this PR

fjetter commented:
I also checked again https://github.com/pydata/xarray/issues/8414 which now produces this ordering

![image](https://github.com/dask/dask/assets/8629629/39d59ebd-da4b-4180-b7c9-f76bb7469d7a)

which I believe is just right. I'll add this as a test case as well.

fjetter commented:
A note about the implementation:

The complexity went through the roof but I think it's worth it. I tried factorizing it but it doesn't break apart easily. However, there are very clear stages and I split off those stages and modularized it as good as possible. I didn't manage to factor this stuff out to functions but I still ended up using local functions to manage scope and namespace pollution.
If anybody is brave enough to review I suggest to start towards the bottom of the `order` function where the core algorithm starts and is documented. I'll eventually follow up with more docs about the other sections.

mrocklin commented:
It looks like the bus factor went from zero to one.

fjetter commented on dask/dataframe/tests/test_groupby.py:
```diff
@@ -838,7 +838,9 @@ def test_split_apply_combine_on_series(empty):
 @pytest.mark.parametrize("keyword", ["split_every", "split_out"])
-def test_groupby_reduction_split(keyword, agg_func):
+def test_groupby_reduction_split(keyword, agg_func, shuffle_method):
+    if agg_func in {"first", "last"} and shuffle_method == "disk":
+        pytest.skip(reason="https://github.com/dask/dask/issues/10034")
```
Apparently, the reordering made us hit https://github.com/dask/dask/issues/10034 in the `test_groupby_reduction_split`

fjetter commented:
Apparently, the reordering made us hit https://github.com/dask/dask/issues/10034 in the `test_groupby_reduction_split`. What's even stranger is that it also causes other failures that are more meta related, e.g. a groupby.shift is losing it's `freq` argument in `test_groupby_shift_with_freq`.

fjetter commented on dask/tests/test_order.py:
```diff
@@ -1236,9 +1237,40 @@ def test_anom_mean():
+    # `mean_chunk` which is the primary reducer in this graph. Therefore we want
+    # to run those as quickly as possible.
+    # This is difficult to assert on but the pressure is an ok-ish proxy
+    assert max(pressure) <= 200
+    transpose_metrics = {k: v for k, v in diags.items() if k[0].startswith("transpose")}
+    assert len(transpose_metrics) == ngroups
+    dependencies, dependents = get_deps(dsk)
+    # This is a pretty tightly connected graph overall and we'll have to hold
+    # many tasks in memory until this can complete. However, we should ensure
+    # that we get to the mean_chunks asap while the transposes are released
+    # quickly.
+    # If this breaks, I suggest to visually inspect the graph and run the above
+    # on a single threaded LocalCluster and verify that the progress is indeed
+    # in five steps (i.e. five groups)
+    ages_mean_chunks = {k: v.age for k, v in diags.items() if "mean_chunk" in k[0]}
+    avg_age_mean_chunks = sum(ages_mean_chunks.values()) / len(ages_mean_chunks)
+    max_age_mean_chunks = max(ages_mean_chunks.values())
+    ages_tranpose = {k: v.age for k, v in diags.items() if k[0].startswith("transpose")}
+    assert max_age_mean_chunks > 1000
+    assert avg_age_mean_chunks > 300
+    avg_age_transpose = sum(ages_tranpose.values()) / len(ages_tranpose)
+    max_age_transpose = max(ages_tranpose.values())
+    assert max_age_transpose < 150
+    assert avg_age_transpose < 100
+    assert sum(pressure) / len(pressure) < 100
```
This one is/was tricky. We have patterns like the following very far into the graph. The `process_runnable` logic should've taken care of those things since those getitem tasks connect very well to other parts of the graph.

However, since this is pretty entangled we have to run the process_runnable until there are no more (unknown) runnables.

![image](https://github.com/dask/dask/assets/8629629/ed59a354-6e3c-4474-992b-832231840d19)

fjetter commented on dask/order.py:
```diff
@@ -115,786 +106,434 @@ def order(
+
+    @_with_offset
+    def process_runnables() -> None:
+        while runnable:
```
This was added after the latest benchmarks ran. While all test are green and some are performing better, I will rerun the coiled benchmarks suite to confirm this isn't causing problems

hendrikmakait replied:
Is there an update on this?

fjetter replied:
https://github.com/coiled/benchmarks/actions/runs/7165432161

Results look pretty much as before. I see again a regression in the test_join cases that I couldn't reproduce locally but I am rerunning the benchmarks for those tests specifically to double check

hendrikmakait commented on dask/tests/test_order.py:
```diff
@@ -1,27 +1,85 @@
 
 
+def visualize(dsk, **kwargs):
+    """Utillity to visualize the raw low level graphs in this tests suite. This
```
```suggestion
    """Utility to visualize the raw low level graphs in this tests suite. This
```

hendrikmakait commented on dask/tests/test_order.py:
```diff
@@ -39,6 +97,7 @@ def test_ordering_keeps_groups_together(abcde):
     assert abs(o[(a, 1)] - o[(a, 3)]) == 1
 
 
+@pytest.mark.skip()
```
Why skip this test?

fjetter replied:
I'll delete this test.

```
    b0    b1  b2
    |      \  /
    a0      a1
```

This topology is a little ambiguous. Earlier versions asserted that we start with `a0` first but from a parallel computing POV it would actually be beneficial to start with `a1` while it has zero impact for serial computing. The only thing we'd want to avoid it to start with either branch and switch to the other one before the branch is done.

I think this test case is too simple to be meaningful

hendrikmakait commented on dask/tests/test_order.py:
```diff
@@ -257,6 +315,7 @@ def test_gh_3055():
 
     dsk = dict(w.__dask_graph__())
     o = order(dsk)
+    assert max(diagnostics(dsk, o=o)[1]) <= 8
```
```suggestion
    # Hold at most 8 outputs in memory at any time
    assert max(diagnostics(dsk, o=o)[1]) <= 8
```

fjetter replied:
I didn't add these suggestions. I don't like comments that spell out what the code already says and I think this is mostly due to unfamiliarity with the `diagnostics` function / pressure metric.

hendrikmakait commented on dask/tests/test_order.py:
```diff
@@ -563,6 +636,8 @@ def test_use_structure_not_keys(abcde):
     }
     o = order(dsk)
+
+    assert max(diagnostics(dsk, o=o)[1]) == 3
```
```suggestion
    # Hold at most 3 outputs in memory at any time
    assert max(diagnostics(dsk, o=o)[1]) == 3
```

hendrikmakait commented on dask/tests/test_order.py:
```diff
@@ -589,7 +664,13 @@ def test_dont_run_all_dependents_too_early(abcde):
         dsk[(d, i)] = (f, (d, i - 1), (b, i), (c, i))
     o = order(dsk)
-
+    # The expected numbers here stand for the central computation branch.
+    # Ideally, they are exactly three apart since we only want to load two nodes
+    # before moving forward with a central node.
+    # This pattern is currently broken for the next to last node since we're
+    # eagerly freeing up the root node before processing once there are no
+    # further dangling dependents. Earlier versions of dask.order behaved
+    # different but this is a net-zero memory pressure operation so it is fine.
```
It looks to me like expected is spaced equally. What exactly is broken here?

fjetter replied:
this is outdated

hendrikmakait commented on dask/tests/test_order.py:
```diff
@@ -1635,36 +1736,292 @@ def test_flaky_array_reduction():
+    assert len(con_r) == len(dsk)
+    assert con_r[(e, 0)] == {("root", 0), (a, 1)}
+    o = order(dsk)
+    assert max(diagnostics(dsk, o=o)[1]) <= 5
```
```suggestion
    # Hold at most 5 outputs in memory at any time  
    assert max(diagnostics(dsk, o=o)[1]) <= 5
```

hendrikmakait reviewed (commented):
A few minor comments/nits from the first superficial round.

hendrikmakait commented on dask/order.py:
```diff
@@ -115,786 +106,427 @@ def order(
+        if not abs(len(root_nodes) - len(leaf_nodes)) / len(root_nodes) > 0.8:
+            # Heavy reducer / splitter topologies often benefit from a very
+            # traditional critical path that expresses the longest chain of
+            # tasks If there are disconnected subgraphs we will only pick a
```
```suggestion
            # tasks. If there are disconnected subgraphs we will only pick a
```

hendrikmakait commented on dask/order.py:
```diff
@@ -115,786 +106,427 @@ def order(
 
-    result: dict[Key, int] = {}
-    i = 0
+    def pick_strategy() -> bool:
```
Nit: I'd suggest naming the strategies and returning their name here. A boolean result of `pick_strategy` is meaningless without additional context.

fjetter replied:
I renamed it to `use_longest_path` to make it explicit. I kind of like the bool version better for now considering that we don't have many strategies and am not totally convinced that I want many

hendrikmakait commented on dask/order.py:
```diff
@@ -115,786 +106,427 @@ def order(
+        # (various metrics, connected roots, etc.) which can be reused for
+        # multiple orderings.
+        size = 0
+        if not abs(len(root_nodes) - len(leaf_nodes)) / len(root_nodes) > 0.8:
```
I'm not sure if this is intended, but the ratios of root and leaf nodes this algorithm considers are very uneven:

For `n_root_nodes := 10`: `abs(10 - n_leaf_nodes) / 10 > 0.8` if `n_leaf_nodes < 2`, i.e., a ratio of more than `5:1`. 
Also, `abs(10 - n_leaf_nodes) / 10 > 0.8` if `n_leaf_nodes > 18`, a ratio of less than `5:9` (or simplifying a bit `1:2`).

To make this more even, we may want to look at `abs(len(root_nodes) - len(leaf_nodes)) / {max|min}(len(root_nodes), len(leaf_nodes))`.

fjetter replied:
This heuristic is a little hand wavy and I can not guarantee that it is correct. This asymmetry is actually accidental but when I tried modifying it I get various test failures that are rightfully false. Just from looking at a couple of examples, it appears as if this kind of asymmetry is beneficial but I believe we do not have sufficient test cases to say for sure.

I'll keep it as is for now

hendrikmakait commented on dask/order.py:
```diff
@@ -115,787 +106,462 @@ def order(
+    dependencies: Mapping[Key, set[Key]], dependents: Mapping[Key, set[Key]]
+) -> tuple[dict[Key, set[Key]], dict[Key, int]]:
+    """Determine for every node which root nodes are connected to it (i.e.
+    ancestors). If arguments of dependencies and depentends are switched, this
```
```suggestion
    ancestors). If arguments of dependencies and dependents are switched, this
```

hendrikmakait commented on dask/order.py:
```diff
@@ -115,787 +106,462 @@ def order(
+    #
+    #   1. Determine the target node by calling `get_target`` and append the
+    #      target to the critical path stack
+    #   2. Take the _most valuable_ (max given a `sort_key`) of it's dependends
```
```suggestion
    #   2. Take the _most valuable_ (max given a `sort_key`) of its dependents
```

hendrikmakait commented on dask/order.py:
```diff
@@ -115,787 +106,462 @@ def order(
+    #   To build the critical path we will use a provided `get_target` function
+    #   that returns a node that is anywhere in the graph, typically a leaf
+    #   node. This node is not required to be runnable. We will walk the graph
+    #   backwards and append nodes to the graph as we go. The critical path is a
```
*backwards* as in "toward the roots"? If so, I'd make this explicit.

fjetter commented:
Re algorithmic performance, this implementation is now slower than before but it look to be reasonable and in real world testing it doesn't seem to have an impact (the other stuff we're doing on the scheduler is just much slower).

Particularly, the dask-benchmark tests I've been running locally to test performance are often hitting the worst cases of this new algorithm.

Overall, I'm not concerned by a (~linear complexity) increase in runtime

hendrikmakait commented on dask/order.py:
```diff
@@ -115,787 +106,516 @@ def order(
-            )
-        for child in dependencies[key]:
+        key = current.pop()
+        for child in dependents[key]:
```
nit: The definition of `parent` and `child` in `_connecting_to_roots` is inverse to that of `ndependencies`. We should align this.

fjetter commented on dask/order.py:
```diff
@@ -115,787 +106,462 @@ def order(
+            if not previous:
+                previous = result[parent]
+                max_dependents[key] = max_dependents[parent]
+            elif identical_sets and (
+                previous is result[parent]
+                or (len(previous) == len(result[parent]) and previous == result[parent])
+            ):
+                identical_sets = True
+            else:
+                identical_sets = False
+                max_dependents[key] = max(max_dependents[parent], max_dependents[key])
+                new_set.update(result[parent])
```
yeah, this evolved over time and I rewrote it already once or twice. I'm fine with your suggestions

fjetter replied:
logically, those expressions are identical but they do observe different performance

consider the case where `identical_sets=True` and `previous is result[parent]` 

My former implementation would evaluate to `not True or True or <expensive-check>` where `<expensive-check>` is the length plus content check, i.e. for the exact same python set/object we would skip the more expensive check for equal elements.

Your implementation otoh would evaluate to

`False or (False and <expensive-check>)`

i.e. we could not benefit from short circuiting. I'll revert this again

fjetter replied:
however, I also notice a bug so I'll rewrite this again

hendrikmakait replied:
> Your implementation otoh would evaluate to
> 
> `False or (False and <expensive-check>)`
> 
> i.e. we could not benefit from short circuiting. I'll revert this again

FWIW, `False or (False and <expensive_check>)` also short-circuits because `False and <whatever>` is always `False`.

fjetter commented:
So, I saw again regressions in set_index benchmarks that somehow pop up on the ordering. I just cannot understand how ordering would have an impact on P2P shuffling. When inspecting the dashboard, two things pop up that probably cause the benchmark failure

![image](https://github.com/dask/dask/assets/8629629/531865c2-1602-4fc9-8ec9-be2afa530e02)


For some reason the unpack tasks are flagged as root tasks and are queued which causes a heavy skew in assigned tasks which causes work stealing and a lot of irrelevant data transfer to compute the `size`. @hendrikmakait the p2p tasks previously carried worker annotations which effectively blocked them from every be flagged rootish. By removing those annotations we might have opened ourselves up to this scheduling anomaly. I don't understand how this connects to ordering and I suspect this is unrelated. Maybe the ordering change changed some subtle timing somewhere that makes it more likely to hit this.

fjetter commented on dask/order.py:
```diff
@@ -115,787 +106,516 @@ def order(
+        cpath_update = scrit_path.update
+        cpath_add = scrit_path.add
+        cpath_discard = scrit_path.discard
+
```
I rewrote this section. Flake8 was actually very useful here since it made me realize that I can reuse the same crticial_path list instead of having it recreated. This add/pop/etc. stuff is now a function that updates both collections as necessary

fjetter commented:
OK, I think I nailed down the last failing test. Will move forward with merging unless there are any objections

fjetter commented on dask/order.py:
```diff
@@ -115,787 +103,525 @@ def order(
+                    result_first = r_child
+                    max_dependents[key] = max_dependents[child]
+                # This clause is written such that it can circuit break early
+                elif not (  # type: ignore[unreachable]
+                    identical_sets
+                    and (result_first is r_child or r_child.issubset(result_first))
+                ):
```
@hendrikmakait I won't hold off merging for your review here but since we talked about it before this is the rewritten version. I think this much cleaner now (and I even wrote a couple of dedicated unit tests 🎉 )

dcherian commented:
This needs a blogpost to get the word out!
